### PR TITLE
if it is a mount point we need to get the path from the mount manager

### LINKED ISSUE #12106: External Storage: Sub folder group sharing broken
When I share a root folder, in this case a folder called WND, a windows network drive, with the group USERS. This group has members user1, 2, 3, 4, 5 and user6

When I, as the owner of the folder, click into the windows network drive, I should then be able to click on share for a sub folder and see "shared WND with group users". Instead, I see a list of the members of the group in sub folders, and the word FALSE instead of the name of the shared folder. See screenshot.
untitled

### PR DESCRIPTION
**Pull request by schiessle**
steps to test:
- create three user: user1, user2, user3
- add user2 and user3 to the group 'users'
- user1 mounts a external storage to "MountPoint"
- user1 uploads a file to the external storage
- user1 shares the mount point with the group 'users'
- user1 navigate to the external storage and opens the share drop-down of the uploaded file

Without this patch you will see:
"Shared in {item} with users"

With this patch:
"Shared in MountPoint with users"

fix https://github.com/owncloud/core/issues/12106

### CODE REVIEW

**Comment by schiesbn** [comment]
@rullzer maybe you have some time to review it? Thanks!

**Comment by nickvergessen** [comment]
:+1:

**Comment by nickvergessen** [comment]
Easy to review and test, @MorrisJobke

**Comment by MorrisJobke** [comment]
Handover to @LukasReschke :P

**Comment by LukasReschke** [comment]
Handsback to @MorrisJobke.

File: lib/private/share/share.php [hunk 1]
```diff
@@ -372,6 +372,15 @@ public static function getItemSharedWithUser($itemType, $itemSource, $user, $own
 			if ($fileDependent && !self::isFileReachable($row['path'], $row['storage_id'])) {
 				continue;
 			}
+			if ($fileDependent && $row['file_parent'] === '-1') {
+				// if it is a mount point we need to get the path from the mount manager
+				$mountManager = \OC\Files\Filesystem::getMountManager();
+				$mountPoint = $mountManager->findByStorageId($row['storage_id']);
+				$path = $mountPoint[0]->getMountPoint();
+				$path = rtrim($path, '/');
```
Comment (PVince81) [inline lib/private/share/share.php]: Might be safer to use `trim($path, '/')`. Never assume that `getMountPoint` will give you something with a leading slash...

Comment (schiesbn) [inline lib/private/share/share.php]: I want to remove the trailing slash, if it exists.

Comment (PVince81) [inline lib/private/share/share.php]: Yes, my point was that you should also remove the leading slash if it exists.

Comment (schiesbn) [inline lib/private/share/share.php]: ah, ok... yes, makes probably sense. I will update the PR

File: .idea/codeStyleSettings.xml [hunk 2]
```diff
@@ -47,6 +47,6 @@
       </value>
     </option>
     <option name="USE_PER_PROJECT_SETTINGS" value="true" />
+    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Default (1)" />
   </component>
-</project>
-
+</project>
```
Comment (schiesbn) [inline .idea/codeStyleSettings.xml]: will push revert once Jenkins finished. Don't know why my IDE always manipulate this value since today...

File: lib/private/share/share.php (hunk 1 above)
Comment (schiesbn) [inline lib/private/share/share.php]: done @PVince81 please have another look, thanks!

File: .idea/codeStyleSettings.xml (hunk 2 above)
Comment (MorrisJobke) [inline .idea/codeStyleSettings.xml]: Welcome in the club ;)

Comment (schiesbn) [inline .idea/codeStyleSettings.xml]: did it happen to you too? Any idea how to solve it? The strange thing is that it only happens since today but I didn't updated anything on my system and the branch already existed yesterday. Unless PHPStorm did some updates this morning in the background and didn't notified me.

Comment (LukasReschke) [inline .idea/codeStyleSettings.xml]: Yup. Please complain at https://github.com/owncloud/core/pull/14287 and write that we want this :speak_no_evil:

File: lib/private/share/share.php [hunk 3]
```diff
@@ -372,6 +372,22 @@ public static function getItemSharedWithUser($itemType, $itemSource, $user, $own
 			if ($fileDependent && !self::isFileReachable($row['path'], $row['storage_id'])) {
 				continue;
 			}
+			if ($fileDependent && $row['file_parent'] === '-1') {
```
Comment (PVince81) [inline lib/private/share/share.php]: -1 string ? Is there a risk that sometimes it will be an int ?

File: lib/private/share/share.php [hunk 4]
```diff
@@ -2290,7 +2306,7 @@ private static function createSelectStatement($format, $fileDependent, $uidOwner
 			if ($fileDependent) {
 				$select = '`*PREFIX*share`.`id`, `*PREFIX*share`.`parent`, `share_type`, `path`, `storage`, '
 					. '`share_with`, `uid_owner` , `file_source`, `stime`, `*PREFIX*share`.`permissions`, '
-					. '`*PREFIX*storages`.`id` AS `storage_id`';
+					. '`*PREFIX*storages`.`id` AS `storage_id`, `*PREFIX*filecache`.`parent` as `file_parent`';
```
Comment (PVince81) [inline lib/private/share/share.php]: I assume you have checked that the join with filecache is available in all code paths leading here

**Comment by PVince81** [comment]
Tested with personal mount, works.

Tested with system mount assigned to only user1, the share dropdown will show "Shared in externalmount with users, user2, user3". Might be a separate issue ?

File: lib/private/share/share.php (hunk 3 above)
Comment (DeepDiver1975) [inline lib/private/share/share.php]: the parent column of filecache is an int - I'd expect this to be compared against -1 instead of '-1'

File: lib/private/share/share.php (hunk 4 above)
Comment (schiesbn) [inline lib/private/share/share.php]: yes, we only do it for file/folder shares, see $fileDependent variable

Comment (PVince81) [inline lib/private/share/share.php]: good, because I don't remember all the code paths :smile:

File: lib/private/share/share.php (hunk 3 above)
Comment (schiesbn) [inline lib/private/share/share.php]: to my knowledge PHP convert all query results to string. If you grep the code you will find places where we explicitly cast it back to int. But the result of a db query always returns all values as string.

Comment (PVince81) [inline lib/private/share/share.php]: Is this covered by a unit test ? It would make sure that all tested DBs would return the proper value.

By the way, this PR seems to have no unit test change ?

**Comment by schiesbn** [comment]
> Tested with system mount assigned to only user1, the share dropdown will show "Shared in externalmount with users, user2, user3". Might be a separate issue ?

I can't reproduce it. Works also with system mount here.

**Comment by schiesbn** [comment]
My steps:


- create three user: user1 (admin), user2, user3
- add user2 and user3 to the group 'users'
- user1 mounts a external storage to "MountPoint" as admin only for user1
- user1 uploads a file to the external storage
- user1 shares the mount point with the group 'users'
- user1 navigate to the external storage and opens the share drop-down of the uploaded file

File: lib/private/share/share.php (hunk 3 above)
Comment (schiesbn) [inline lib/private/share/share.php]: > By the way, this PR seems to have no unit test change ?

We don't really have a way to test this explicitly because of all the static classes and methods.

Comment (PVince81) [inline lib/private/share/share.php]: Can't wait for the full rewrite.

Fine by me.

Comment (schiesbn) [inline lib/private/share/share.php]: As said, that's the way it is done all over the code and if you search the internet you will also find various discussions where people always explain that this is how PHP works. But I changed the code for now to cast it to a int and then compare the int values, just to give us all a better feeling :wink:

**Comment by schiesbn** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
These were my steps, yes. Strange.
Let me try again from scratch.

**Comment by PVince81** [comment]
I retried the steps and it works now.
I wonder if it's because I initially had the global mount applied to all and only later set it to "user1".

Anyway, fine now :+1:

**Comment by davitol** [comment]
:+1: It works in 8.1 RC2
